src/hashstore/hashstore.py

```diff
@@ -1,103 +1,708 @@
+        tmp = NamedTemporaryFile(dir=tmp_root_path, delete=False)
+
+        # Ensure tmp file is created with desired permissions
+        if self.fmode is not None:
+            oldmask = os.umask(0)
+            try:
+                os.chmod(tmp.name, self.fmode)
+            finally:
+                os.umask(oldmask)
+
+        # Additional hash object to digest
+        if algorithm is not None:
+            if algorithm in self.other_algo_list:
+                self.default_algo_list.append(algorithm)
+            elif algorithm not in self.default_algo_list:
+                raise ValueError(f"Algorithm not supported: {algorithm}")
+
+        hash_algorithms = [
+            hashlib.new(algorithm) for algorithm in self.default_algo_list
+        ]
+
+        # tmp is a file-like object that is already opened for writing by default
+        with tmp as tmp_file:
+            for data in stream:
+                tmp_file.write(self._to_bytes(data))
+                for hash_algorithm in hash_algorithms:
+                    hash_algorithm.update(self._to_bytes(data))
+
+        hex_digest_list = [
+            hash_algorithm.hexdigest() for hash_algorithm in hash_algorithms
+        ]
+        hex_digest_dict = dict(zip(self.default_algo_list, hex_digest_list))
+
+        return hex_digest_dict, tmp.name
+
+    def put_sysmeta(self, pid, sysmeta, namespace):
+        """Store contents of `sysmeta` on disk using the hash of the given pid
+
+        Args:
+            pid (string): authority-based identifier
+            sysmeta (mixed): string or path to sysmeta document
+            namespace (string): sysmeta format
+
+        Returns:
+            ab_id (string): address of the sysmeta document
+        """
+
+        # Create tmp file and write to it
+        sysmeta_stream = Stream(sysmeta)
+        with closing(sysmeta_stream):
+            sysmeta_tmp = self._mktmpsysmeta(sysmeta_stream, namespace)
+
+        # Target path (permanent location)
+        ab_id = self.get_sha256_hex_digest(pid)
+        rel_path = "/".join(self.shard(ab_id))
+        full_path = self._get_store_path() / rel_path
+
+        # Move sysmeta to target path
+        if os.path.exists(sysmeta_tmp):
+            try:
+                parent = full_path.parent
+                parent.mkdir(parents=True, exist_ok=True)
+                # Sysmeta will be replaced if it exists
+                shutil.move(sysmeta_tmp, full_path)
+                return ab_id
+            except Exception as err:
+                # TODO: Discuss specific error handling
+                # isADirectoryError/notADirectoryError - if src/dst are directories, cannot move
+                # OSError - if dst is a non-empty directory and insufficient permissions
+                # TODO: Log error - err
+                # Delete tmp file if it exists
+                if os.path.exists(sysmeta_tmp):
+                    self.sysmeta.delete(sysmeta_tmp)
```

Comment:
The `self` here refers to the current sysmeta file store instance because of the way `HashStore` python is currently initialized, one `HashFSExt` instance each for object and sysmeta. This is not the intended final behaviour, and will be addressed next in [issue-23 - Replace HashFS with new class FileHashStore](https://github.com/DataONEorg/hashstore/issues/23)